### PR TITLE
[StorageQuota] Add some quota interfaces and a helix storage usage implementation

### PR DESCRIPTION
**Pull request by justinlin-linkedin**
This PR
1. Add several interfaces for storage quota in frontend
2. Add an implementation that fetch and refresh storage usage through HelixPropertyStore.

### CODE REVIEW

**Review by jsjtzyy — COMMENTED** [review]
LGTM, after minor comments are addressed. 
Looks like ambry-quota is not a module in Intellj when I pulled the changes to my local repo.

File: ambry-api/src/main/java/com/github/ambry/quota/StorageQuotaService.java
Comment (justinlin-linkedin) [reply ambry-api/src/main/java/com/github/ambry/quota/StorageQuotaService.java:16]: Will add doc later when I add methods to this interface.

File: ambry-quota/src/main/java/com/github/ambry/quota/HelixStorageUsageRefresher.java
Comment (justinlin-linkedin) [reply ambry-quota/src/main/java/com/github/ambry/quota/HelixStorageUsageRefresher.java:47]: This is not configurable. Whatever goes to config would be considered a configurable value. I am thinking about abstract some of the logic in this class out with aggregation task since they share the same znode path and the same format.

File: ambry-quota/src/main/java/com/github/ambry/quota/HelixStorageUsageRefresher.java
Comment (justinlin-linkedin) [reply ambry-quota/src/main/java/com/github/ambry/quota/HelixStorageUsageRefresher.java:84]: that's right. I don't know whether helix would fetch the content of z path if we call helix.subscribe so here I just blindly make the assumption that in memory cache might not be null.

File: ambry-api/src/main/java/com/github/ambry/quota/StorageUsageRefresher.java
Comment (justinlin-linkedin) [reply ambry-api/src/main/java/com/github/ambry/quota/StorageUsageRefresher.java:40]: Will change the name of the interface.

File: ambry-quota/src/main/java/com/github/ambry/quota/JSONStringStorageQuotaSource.java
Comment (justinlin-linkedin) [reply ambry-quota/src/main/java/com/github/ambry/quota/JSONStringStorageQuotaSource.java:46]: This is for open source implementation, the closed source would be implemented in ambryli.

**Comment by justinlin-linkedin** [comment]
> LGTM, after minor comments are addressed.
> Looks like ambry-quota is not a module in Intellj when I pulled the changes to my local repo.

Do you know how to set this to be a module in intellj?

File: ambry-api/src/main/java/com/github/ambry/quota/StorageQuotaService.java
Comment (ankagrawal) [reply ambry-api/src/main/java/com/github/ambry/quota/StorageQuotaService.java:16]: Might be nitpicking here, so I will let you and @jsjtzyy take the final decision on this. But I am not sure why are you even adding the interfaces if there is no method in that now? They don't serve any purpose for now. I think you can keep the three new interfaces StorageQuotaService.java, StorageQuotaServiceFactory.java and StorageQuotaEnforcer.java in a separate PR where you have figured out their behavior and methods.

File: ambry-api/src/main/java/com/github/ambry/quota/StorageUsageRefresher.java
Comment (ankagrawal) [inline ambry-api/src/main/java/com/github/ambry/quota/StorageUsageRefresher.java:20]: Sorry about the nitpick again, but I am not sure what do we mean by refresh here? Does it mean refreshing a cache, or a backend. Or does it mean updating something with new information?

File: ambry-api/src/main/java/com/github/ambry/quota/StorageUsageRefresher.java
Comment (ankagrawal) [inline ambry-api/src/main/java/com/github/ambry/quota/StorageUsageRefresher.java:32]: This is more like Account Storage Usage. Because you are basically getting quota usage for all containers of an account.

File: ambry-quota/src/main/java/com/github/ambry/quota/HelixStorageUsageRefresher.java
Comment (ankagrawal) [inline ambry-quota/src/main/java/com/github/ambry/quota/HelixStorageUsageRefresher.java:131]: If only one listener can ever be registered, can we pass it in the constructor?

File: ambry-quota/src/main/java/com/github/ambry/quota/HelixStorageUsageRefresher.java
Comment (ankagrawal) [inline ambry-quota/src/main/java/com/github/ambry/quota/HelixStorageUsageRefresher.java:39]: Are we not persisting the current storage usage in zookeeper now?

**Review by ankagrawal — COMMENTED** [review]
Left some initial comments. Will take one more look at it today.

File: ambry-api/src/main/java/com/github/ambry/quota/StorageQuotaService.java
Comment (justinlin-linkedin) [reply ambry-api/src/main/java/com/github/ambry/quota/StorageQuotaService.java:16]: I will remove those interfaces then.

File: ambry-api/src/main/java/com/github/ambry/quota/StorageUsageRefresher.java
Comment (justinlin-linkedin) [reply ambry-api/src/main/java/com/github/ambry/quota/StorageUsageRefresher.java:20]: It means if the storage usage is changed, by any other outsider, the implementation should be able to re-fetch the storage usage and refresh the in memory cache.

File: ambry-api/src/main/java/com/github/ambry/quota/StorageUsageRefresher.java
Comment (justinlin-linkedin) [reply ambry-api/src/main/java/com/github/ambry/quota/StorageUsageRefresher.java:32]: It actually returns container storage usage for all accounts. The outer map's key is the account id, so it covers all the accounts and containers. The reason why this is "getContainerStorageUsage" not "getAccountStorageUsage" is to emphasize the storage usage is per container, not per account.

File: ambry-quota/src/main/java/com/github/ambry/quota/HelixStorageUsageRefresher.java
Comment (justinlin-linkedin) [reply ambry-quota/src/main/java/com/github/ambry/quota/HelixStorageUsageRefresher.java:131]: It can be passed in the constructor. Basically the refresher interface will be used in StorageQuotaService, which will be implemented later. If passing to constructor makes more sense, i will change it later when implementing quota service.

File: ambry-quota/src/main/java/com/github/ambry/quota/HelixStorageUsageRefresher.java
Comment (justinlin-linkedin) [reply ambry-quota/src/main/java/com/github/ambry/quota/HelixStorageUsageRefresher.java:39]: They are in ZooKeeper. I am using Helix client to fetch storage usage from zookeeper.

Comment (ankagrawal) [reply ambry-quota/src/main/java/com/github/ambry/quota/HelixStorageUsageRefresher.java:39]: Ok. I was asking because in the comments here we are saying "A Helix aggregation task, created in ambry-server, would persist current storage usage in ZooKeeper periodically." Should the persist part be removed from the comment?

**Review by ankagrawal — APPROVED** [review]
looks good to me after one minor change in comments.

**Review by jsjtzyy — APPROVED** [review]

**Comment by justinlin-linkedin** [comment]
> Ok. I was asking because in the comments here we are saying "A Helix aggregation task, created in ambry-server, would persist current storage usage in ZooKeeper periodically." Should the persist part be removed from the comment?

This is just more context for the implementation. I don't think we need to remove it.

**Pull request merged**
